hedera-node/docs/design/services/smart-contract-service/hedera-account-service-isAuthorized-design.md

```diff
@@ -0,0 +1,88 @@
+[HIP-632](https://hips.hedera.com/hip/hip-632) - "Hedera Account Service (HAS) 
+System Contract".
+
+## Preliminary notice w.r.t. "virtual addresses"
+
+HIP-632 as written refers a lot to 
+[HIP-631](https://hips.hedera.com/hip/hip-631) - "Account Virtual Addresses".
+However, that HIP, though `Accepted` will not be implemented.  (Or at least, not
+at any time soon.)  So -632 needs to be read with that in mind.  Specifically, 
+anywhere that there's a provision for dealing with "virtual addresses", of which there
+may be several/many per Hedera account, there is in fact only _one_: the _alias_.
+
+* This mainly affects the method `getVirtualAddresses` which will return an array, but
+it will have _at most one_ EVM address in it.
+  
+## Phased implementation
+
+The first release of -631 will include only `isVirtualAddress` and `isAuthorizedRaw`.
+
+## Where does this go?
+
+A "hedera account service" system contract is being set up in
+`com.hedera.node.app.service.contract.impl.exec.systemcontracts.has`, and also at
+`...contract.impl.exec.processors.HasTranslatorsModule` (for the individual methods).
+
+Following the existing pattern for HAS each method will be given a package under 
+`...contract.impl.exec.systemcontracts.has` with two classes:
+* `[method]Translator` - which first confirms that the call is for the given method, and
+  then builds a `TransactionBody` for it.
+* `[method]Call` - which takes that `TransactionBody` and then has an `execute()` method
+  that does the work.
+
+
+## Design for `isVirtualAddress`
+
+Looking up a Hedera account from an EVM address is done through the
+`HederaWorldUpdator.Enhancement` which can map from an EVM address to the Hedera account it
+is an alias for.  This functionality is provided by `HasCallAttempt.linkedAccount()`.
+
+## Design for `isAuthorizedRaw`
+
+[TBD shortly as I fix up some problems with understanding HAS]
+
+`isAuthorizedRaw` has two flows: Either the argument address is a Hedera account, or it is
+an EVM address.
+
+* If an EVM address then this is simply the `ECRECOVER` precompile.  We have direct
+access to the `ECRECOVER` precompile from BESU in the class
+`org.hyperledger.besu.evm.precompile.ECRECPrecompiledContract` which will compute
+the `ECRECOVER` operation given the message hash + signature (encoded as a calldata for
+the precompile) (and a `MessageFrame` which it ignores).  This returns either the
+recovered address, if successful, or "empty", if not.  The recovered address must then
+match the address given as an argument to `isAuthorizedRaw`.
+  
+* If an Hedera account then it gets the account, looks in it to find a key _which it must
+have_ and which must be a _single key_ (otherwise: failure).  Then given the key and the
+signature verifies that the signature matches the message hash and is attested by 
+the account. This is an operation provided by `HandleContext` (which can be accessed via
```

Comment:
A little more detail would be nice (which methods are needed from `HandleContext` and what information do they provide.

Reply:
Will do.